### ISSUE TITLE
File descriptors are leaked when using HTTP

### ISSUE THREAD

**Issue by ryanbrink** [elastic/logstash#1604]
Several of Logstash's input and output modules use HTTP via the FTW gem (https://github.com/jordansissel/ruby-ftw). From my experience with Logstash, this gem is leaking file descriptors and eventually causing Logstash to essentially die by repeatedly trying and failing to open a file descriptor for some task. For example, here is a trace I got after running Logstash for a couple of days in my setup:

```
{:timestamp=>"2014-08-06T11:13:07.128000+0000", 
:message=>"Failed to flush outgoing items", 
:outgoing_count=>1, 
:exception=>#<SocketError: initialize: java.net.SocketException: Too many open files>, 
:backtrace=>[
    "org/jruby/ext/socket/RubySocket.java:190:in `initialize'", 
    "org/jruby/RubyIO.java:852:in `new'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/connection.rb:146:in `connect'", 
    "org/jruby/RubyArray.java:1613:in `each'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/connection.rb:139:in `connect'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/agent.rb:406:in `connect'", 
    "org/jruby/RubyProc.java:271:in `call'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/pool.rb:48:in `fetch'", "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/agent.rb:403:in `connect'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/agent.rb:319:in `execute'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/ftw-0.0.39/lib/ftw/agent.rb:217:in `post!'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/outputs/elasticsearch_http.rb:215:in `post'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/outputs/elasticsearch_http.rb:210:in `flush'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:219:in `buffer_flush'", 
    "org/jruby/RubyHash.java:1339:in `each'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:216:in `buffer_flush'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:193:in `buffer_flush'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:159:in `buffer_receive'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/outputs/elasticsearch_http.rb:188:in `receive'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/outputs/base.rb:86:in `handle'", 
    "(eval):335:in `initialize'", 
    "org/jruby/RubyProc.java:271:in `call'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/pipeline.rb:266:in `output'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/pipeline.rb:225:in `outputworker'", 
    "/usr/local/liveview/shipping/logstash-1.4.2/lib/logstash/pipeline.rb:152:in `start_outputs'"
    ], 
:level=>:warn}
```

This will repeat until I restart the process, after which Logstash will run ok until it again eats up all available file descriptors on the system. 

You can check the file descriptors in use by running `sudo lsof | grep java | wc -l`. On a busy system (like mine) Logstash is leaking about 10 file descriptors a minute. The number goes up and down as some are released back into the wild, but grows consistently over time.

This issue seems similar to https://logstash.jira.com/browse/LOGSTASH-892.

**tl;dr: FTW is leaking file descriptors, causing any and all HTTP-based plugins to eventually stall Logstash**

_Note: this error can also manifest as a DNS lookup failure (`#<SocketError: gethostbyname: name or service not known>,`), but I believe this is also fundamentally due to a lack of available file descriptors._

**Comment by vanga** [comment]
I have updated the two files mentioned here https://github.com/jordansissel/ruby-ftw/commit/b065c3c7a4463d04a2e8107510625f59eccc18a3
I still dont see the issue fixed, file descriptors goes to max value.
I am using logstast-1.4.2
Is there anyway for me to use ftw 0.0.40 version without having to update logstash?

**Comment by vasubesimple** [comment]
{:timestamp=>"2015-03-24T13:41:07.097000+0000", :message=>"Failed to flush outgoing items", :outgoing_count=>1, :exception=>java.lang.OutOfMemoryError: Java heap space, :backtrace=>[], :level=>:warn}


I have 103 Gigs of free memory !  I am using 1.4.2-1-2c0f5a1 Version !!!! It seems issue was not solved  :) any recommendations ?

**Comment by chiriaev** [comment]
Guys, Hi. I can also see that on my "1.5 RC2". I currently have 3500+ open file descriptors on my log-stash TCP input, port 12201, for about 400 real servers. (i.e.: the value I have after a re-start)

Also, I have a case when my TCP input crashed. (i.e.: as per the "[B.]" output below)

Happy to provide you with more information. Please 

Regards, Andrei

P.S.: again "[A.]" is for an working/running instance with a socket leak
                  "[B.]" is for a similar "+" the TCP input crashed.

+++
[A.]
...
:timestamp=>"2015-03-23T20:21:40.316000+0000", :message=>"An error occurred. Closing connection", :client=>"10.51.4.5:59480", :exception=>#<TypeError: can't convert Array into String>, :backtrace=>["org/jruby/RubyString.java:1156:in `+'", "/opt/logstash/lib/logstash/inputs/base.rb:114:in `decorate'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:119:in `handle_socket'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-json_lines-0.1.6/lib/logstash/codecs/json_lines.rb:40:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-json_lines-0.1.6/lib/logstash/codecs/json_lines.rb:37:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-line-0.1.5/lib/logstash/codecs/line.rb:36:in `decode'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-line-0.1.5/lib/logstash/codecs/line.rb:35:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-json_lines-0.1.6/lib/logstash/codecs/json_lines.rb:35:in `decode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:116:in `handle_socket'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:145:in `client_thread'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:143:in `client_thread'"], :level=>:error}
...

[B]
...
{:timestamp=>"2015-03-24T14:47:15.764000+0000", :message=>"An error occurred. Closing connection", :client=>"10.51.4.5:33399", :exception=>#<IOError: Too many open files>, :backtrace=>["org/jruby/RubyIO.java:2994:in `sysread'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:161:in `read'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:110:in `handle_socket'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:145:in `client_thread'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-tcp-0.1.3/lib/logstash/inputs/tcp.rb:143:in `client_thread'"], :level=>:error}
{:timestamp=>"2015-03-24T14:47:16.837000+0000", :message=>"A plugin had an unrecoverable error. Will restart this plugin.\n  Plugin: <LogStash::Inputs::Tcp type=>\"nxlog_json\", tags=>[\"input_nxlog_json\"], host=>\"0.0.0.0\", mode=>\"server\">\n  Error: Too many open files", :level=>:error}
...

**Comment by f-ld** [comment]
Hi,

I am not using the http output plugin but I had a look to it to make a kind of http input plugin and noticed the same "descriptor leak" issue with ftw.

I made a "quick'n'dirty" workaround to my input plugin by adding a "@agent.shutdown" after my request and it works as expected.

FYI, the agent.shutdown is coded like this
```ruby
# shutdown this agent.
#
# This will shutdown all active connections.
def shutdown
    @pool.each do |identifier, list|
      list.each do |connection|
        connection.disconnect("stopping agent")
      end
    end
  end # def shutdown 
```
So this is basically forcing ftw to close all the connections it has. 
Fixing the leak first, optimization later.
No change needed in the ftw source code.

I guess you can do the same thing inside the http output plugin, at the end of the receive method (just before "end # def receive"). Since I am not using this plugin, I cannot test it very quickly and see if this helps.

For me this is only a workaround but avoids our logstash instances to crash.

Cheers,
Fabien
